holoviews/core/util.py

```diff
@@ -24,6 +26,57 @@
 except ImportError:
     dd = None
 
+
+
+
+class HashGenerator(json.JSONEncoder):
+    """
+    Extends JSONEncoder to generate a hashable string for as many types
+    of key as possible. These keys are supplied by widgets and Stream
+    parameters and may be nested. If an unrecognized object type is
+    supplied, the objects id is used - this may be an issue when dealing
+    with mutable objects. For this reason, it is important to support
+    common mutable types such as pandas Dataframes and numpy arrays
+    (supported).
```

Comment:
The distinction is between what is supported directly by JSON (dicts, lists) and things that we need to add support for it ``HashGenerator``. Currently support is added for sets, datetime, numpy arrays and pandas dataframes/series. More things may need to be explicitly supported in future.